### PR TITLE
#163446117 Voting functionality

### PR DESCRIPTION
**Pull request by oldmonad**

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -29,7 +29,7 @@ app.all('/api/v1', (req, res) => {
 
 app.all('/*', (req, res) => res.status(404).json({
   status: 404,
-  message: 'This route does not exist you might want to check your route specification on postman.',
+  message: 'This route does not exist you might want to check your route specification on postman or check your route configurations.',
```
Comment (Veraclins) [inline server/app.js]: Like I have always told you, you are designing for the end user who is mostly non-technical. Do you think everybody knows what is route let alone postman? Have you visited any site before and received this kind of error message? And do you intend to change this when you implement the frontend or will I be checking postman on the frontend too?

File: server/controller/questions.js [hunk 2]
```diff
@@ -13,14 +13,64 @@ export default {
...
+    }
+
+    const voted = await QuestionModel.ifVoted(userId, questionId);
+
+    if (voted) {
+      return errorResponse(res, 400, `You have already ${voted.vote} this question`);
+    }
+
+    const userActionArray = userAction.split('/');
+    const vote = `${userActionArray[userActionArray.length - 1]}d`;
```
Comment (Veraclins) [inline server/controller/questions.js]: Can you explain what is happening here and why this is necessary?

File: server/models/question.js [hunk 3]
```diff
@@ -39,15 +38,15 @@ class Question {
...
+    const queryPlaceholder = 'UPDATE questions SET down_votes = down_votes + 1 WHERE id = $1 RETURNING *';
     const queryValues = [id];
     const {
       rows,
     } = await pool.query(queryPlaceholder, queryValues);
     return rows[0];
   }
 
-  static async updateVotesTable(userId, questionId, vote) {
+  static async createVotesTable(userId, questionId, vote) {
```
Comment (Veraclins) [inline server/models/question.js]: Why is the name createVotesTable? Are you creating the table here?

File: server/controller/questions.js [hunk 4]
```diff
@@ -13,14 +13,64 @@ export default {
...
+    const userAction = req.originalUrl;
+
+    const questionExists = await QuestionModel.getQuestionById(questionId);
+    if (!questionExists) return errorResponse(res, 404, 'Question not found.');
+
+    if (questionExists.user_id === userId) {
+      return errorResponse(res, 400, 'You can not vote on your own question.');
+    }
+
+    const voted = await QuestionModel.ifVoted(userId, questionId);
```
Comment (Veraclins) [inline server/controller/questions.js]: Does this mean I can't change my vote? What if I wanted to upvote and mistakenly downvote a question?

File: server/controller/rsvp.js [hunk 5]
```diff
@@ -12,26 +16,27 @@ const rsvpController = {
...
+    return successResponse(res, 200, 'Your response has been recorded', null);
+  },
+
+  async getJoinedMeetups(req, res) {
+    const {
+      id,
+    } = req.user;
+    const response = 'yes';
+    const result = await RsvpModels.userResponses(id, response);
+    if (result.length === 0) return successResponse(res, 200, 'You have not joined any meetups yet.', result);
```
Comment (Veraclins) [inline server/controller/rsvp.js]: Is this a successful request?

**Review by Veraclins — CHANGES_REQUESTED** [review]
@dbytecoderc Good job as usual. Just a few points of note for which I have added comments. Let me know what you think about them. Another thing, you need to pay attention to your method names. As a general rule, your method names should say what they do e.g. `upvoteQuestion` but what does `responseToMeetup` do? Also, attaching the Module name to the methods in the module is an unnecessary duplication. `upvoteQuestion` could simply be `upvote`. The Question class would obviously contain only methods related to the Question model. This way when calling them, you can simple do:
`Question.upvote` instead of `Question.upvoteQuestion`
`Question.getById` instead of `Question.getQuestionById`
etc.

File: server/controller/questions.js (hunk 2 above)
Comment (oldmonad) [reply server/controller/questions.js]: I extracted the request URL, converted it to an array and assigned the array to the userActionArray variable then I proceeded to getting the last value which is 'upvote' or 'downvote' and concatenated the letter 'd' to it to make it a past tense before creating a table containing the user id, question id and the 'past tense' i created, So the whole idea is to create a row on that votes table containing the user id and question  as a unique occurrence, the rubric didn't specify if a user can change his/her vote status after voting although I planned implementing that, If you check my questions model you would see I have some methods that is supposed to help with that functionality but at the time of writing this I have not yet figured it out

File: server/controller/rsvp.js (hunk 5 above)
Comment (oldmonad) [reply server/controller/rsvp.js]: It is a successful request but I should have sent a 204 status code response to signify a no content response

File: server/controller/questions.js (hunk 4 above)
Comment (oldmonad) [reply server/controller/questions.js]: the rubric didn't specify if a user can change his/her vote status after voting although I planned implementing that, If you check my questions model you would see I have some methods that is supposed to help with that functionality but at the time of writing this I have not yet figured it out

File: server/models/question.js (hunk 3 above)
Comment (oldmonad) [reply server/models/question.js]: Yeah this table is supposed to serve as a reference point to check the user vote status, at which point I am supposed to update the status to enable the user make another action, for instance if a user has upvoted and wants to downvote I would check this table to know his status and proceed to downvote the question, if he has upvoted and wants to upvote again then i would let him know that he cant upvote twice

File: server/controller/questions.js (hunk 4 above)
Comment (Veraclins) [reply server/controller/questions.js]: You need to own the product and make it as realistic as possible. The rubrics just gives you the bare minimums, it is on you then to provide features that will improve user experience and I think that allowing users the ability to change their vote is cool to have.

File: server/controller/questions.js (hunk 2 above)
Comment (Veraclins) [reply server/controller/questions.js]: I figured but do you think that is necessary? What stops you from just passing the string `upvoted` here and `downvoted` in the other method. In programming, the simplest way is always the best way. Will there ever be a time when this value is noted upvoted?

File: server/controller/rsvp.js (hunk 5 above)
Comment (Veraclins) [reply server/controller/rsvp.js]: How is a request where I didn't get what I requested for a successful request? You must stop thinking from the perspective of the application or developer and start thinking from the perspective of the user. A request is only successful if I get exactly what I requested for.

File: server/controller/questions.js (hunk 2 above)
Comment (oldmonad) [reply server/controller/questions.js]: Ok i think I know a way to make it simpler

File: server/models/question.js (hunk 3 above)
Comment (Veraclins) [reply server/models/question.js]: I beg to disagree, there is a difference between creating a TABLE and creating a RECORD in a table. I believe the latter is the case here. `createVoteTable` cannot be the right method name here.

Comment (oldmonad) [reply server/models/question.js]: Ooh that's true it skipped my mind totally, I would change that to `createVoteRecord` then

**Pull request merged**